openshift/odo
# [Local] Create docker volumes from volumes defined in devfile

**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

/kind feature
/area devfile
/area local

**What does does this PR do / why we need it**:
This PR implements a storage adapter (based off of the Kube storage adapter code) for odo when deploying components to Docker. When deploying or updating local Docker containers, odo determines the unique number of volumes in the devfile and creates Docker volumes for them. Then, when odo is deploying each container, it determines which volumes need to be mounted to it and where. When determining if a container needs to be updated, odo will compare the volumes for the container in the devfile against the volumes mounted in the container, if any differ, odo will update the container with the new (or modified) volume.

Unit tests provided for the new storage adapter, as well as new integration test scenarios to verify the volume gets created and is mounted to the container.

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/2849

**How to test changes / Special notes to the reviewer**:
- Run `make test` and `make test-cmd-docker-devfile-push` to verify the unit and integration tests pass (if running integration tests, make sure pushtarget set to Docker and Docker is installed)

Manual verification:
1. Enable experimental mode and set pushtarget to Docker
2. Run `odo create java-spring-boot`
3. Run `odo push`
4. Verify that the containers deploy successfully and that a "springbootpvc" volume exists (`docker volume ls`)
5. Exec into one of the containers that mount the volume and verify the mount path is there.
